### PR TITLE
Add Versionstamp and Subspace support to Swift bindings

### PR DESCRIPTION
**Pull request by 1amageek**
## Summary

This PR adds two major missing features to the Swift bindings:
1. **Versionstamp support** - Full tuple-layer versionstamp implementation
2. **Subspace support** - Key namespace management with range operations

These features bring the Swift bindings to parity with Python, Go, and Java bindings, enabling the development of Record Layer and other advanced FoundationDB applications in Swift.

## Motivation

The Swift bindings currently lack essential features available in other language bindings, making it difficult to:
- Use versionstamped keys for temporal ordering
- Manage key namespaces efficiently
- Build Record Layer or similar frameworks

This PR addresses these gaps by implementing the missing features following the canonical behavior of official bindings.

### Core Features
- **`Versionstamp` struct**: 12-byte value (10-byte transaction version + 2-byte user version)
- **Incomplete versionstamps**: Support for transaction-time assignment
- **Tuple integration**: Full encode/decode support with type code 0x33
- **`packWithVersionstamp()`**: Automatic offset calculation for atomic operations

### Usage Example
```swift
// Create incomplete versionstamp
let vs = Versionstamp.incomplete(userVersion: 0)
let tuple = Tuple("user", 12345, vs)

// Pack with automatic offset calculation
let key = try tuple.packWithVersionstamp()

// Use with atomic operation
transaction.atomicOp(
    key: key,
    param: value,
    mutationType: .setVersionstampedKey
)

// Read back and decode
let storedKey = try await transaction.get(someKey)
let decoded = try Tuple.decode(from: storedKey)
let versionstamp = decoded[2] as? Versionstamp
```

### Core Features
- **`Subspace` struct**: Key namespace management with tuple encoding
- **`range()` method**: Returns `(prefix + [0x00], prefix + [0xFF])` for tuple data
- **`strinc()` algorithm**: String increment for raw binary prefixes
- **`prefixRange()` method**: Complete prefix coverage using strinc

### Usage Example
```swift
// Create subspace
let userSpace = Subspace(rootPrefix: "users")
let activeUsers = userSpace.subspace("active")

// Pack keys with prefix
let key = activeUsers.pack(Tuple(userId, "name"))

// Range queries
let (begin, end) = activeUsers.range()
let records = try await transaction.getRange(beginKey: begin, endKey: end)

// Raw binary prefix support
let rawSubspace = Subspace(prefix: [0x01, 0xFF])
let (begin, end) = try rawSubspace.prefixRange()  // Uses strinc
```

## Testing

Comprehensive test suite with **150 tests** (all passing):

### Versionstamp Tests (20)
- Incomplete/complete versionstamp creation
- Byte encoding/decoding
- Tuple packing with offset calculation
- **Roundtrip tests** (encode → decode)
- Error handling

### Subspace Tests (22)
- Range operations (`range()` and `prefixRange()`)
- Tuple packing/unpacking
- Namespace containment checks
- Edge cases (0xFF handling)

### String Increment Tests (14)
- strinc algorithm correctness
- Trailing 0xFF handling
- Cross-language compatibility (Java, Go)
- Error cases

### Integration
- All tests verified with Swift Testing framework
- Clean build with no warnings
- Memory-safe operations

### Cross-Language Consistency
This implementation follows the canonical behavior of official bindings:
- **Java**: `ByteArrayUtil.strinc()`, `Versionstamp`, `Subspace`
- **Python**: `fdb.strinc()`, `fdb.tuple.Versionstamp`, tuple packing
- **Go**: `fdb.Strinc()`, `fdb.IncompleteVersionstamp()`, `Subspace.FDBRangeKeys()`
- **C++**: Range semantics match C++ implementation

### API Version
- Supports API 520+ (4-byte offsets)
- Dead code for API < 520 removed for clarity
- Follows modern FoundationDB best practices

### New Files
- `Sources/FoundationDB/Versionstamp.swift` (196 lines)
  - Core Versionstamp implementation
  - TupleElement conformance
  - Comprehensive documentation

- `Sources/FoundationDB/Tuple+Versionstamp.swift` (205 lines)
  - `packWithVersionstamp()` method
  - Validation and helper methods
  - API 520+ offset handling

- `Sources/FoundationDB/Subspace.swift` (424 lines)
  - Subspace implementation
  - `range()` and `prefixRange()` methods
  - `strinc()` algorithm
  - SubspaceError enum

- `Tests/FoundationDBTests/VersionstampTests.swift` (416 lines)
- `Tests/FoundationDBTests/SubspaceTests.swift` (312 lines)
- `Tests/FoundationDBTests/StringIncrementTests.swift` (194 lines)

### Modified Files
- `Sources/FoundationDB/Tuple.swift` (+3 lines)
  - Add versionstamp case to `Tuple.decode()` switch
  - Enables automatic versionstamp decoding

### Statistics
- **Total additions**: ~1,750 lines
- **Test coverage**: 56 new tests
- **Documentation**: Comprehensive inline documentation for all public APIs

## Breaking Changes

None. This PR is purely additive and maintains full backward compatibility with existing code.

## Future Work

This PR lays the groundwork for:
- FoundationDB Record Layer implementation in Swift
- Directory Layer support
- Advanced indexing patterns
- Versionstamp-based temporal queries

## Checklist

- [x] All tests pass (150/150)
- [x] No build warnings
- [x] Comprehensive documentation
- [x] Cross-language compatibility verified
- [x] Memory-safe implementations
- [x] Follows Swift API design guidelines
- [x] Backward compatible

## References

- [FoundationDB Versionstamp Documentation](https://apple.github.io/foundationdb/developer-guide.html#versionstamps)
- [Subspace Documentation](https://apple.github.io/foundationdb/developer-guide.html#subspaces)
- Python bindings: `bindings/python/fdb/tuple.py`
- Go bindings: `bindings/go/src/fdb/tuple/tuple.go`
- Java bindings: `bindings/java/src/main/com/apple/foundationdb/tuple/Versionstamp.java`

### CODE REVIEW

File: Sources/FoundationDB/Subspace.swift [hunk 1]
```diff
@@ -0,0 +1,424 @@
...
+    /// - Throws: `TupleError.invalidDecoding` if the key doesn't start with this prefix
+    ///
+    /// This operation is the inverse of `pack(_:)`. It removes the subspace prefix
+    /// and decodes the remaining bytes as a tuple.
+    public func unpack(_ key: FDB.Bytes) throws -> Tuple {
+        guard key.starts(with: prefix) else {
+            throw TupleError.invalidDecoding("Key does not match subspace prefix")
+        }
+        let tupleBytes = Array(key.dropFirst(prefix.count))
+        let elements = try Tuple.decode(from: tupleBytes)
```
Comment (glbrntt) [inline Sources/FoundationDB/Subspace.swift]: This is quite inefficient: You're allocating a new array and then copying all but one element from the `key` array. It seems like the various `decode` methods might be better implemented as being generic over a `Collection<UInt8>`. That would allow you to decode a slice without going via an intermediate array.

File: Sources/FoundationDB/Subspace.swift [hunk 2]
```diff
@@ -0,0 +1,424 @@
...
+    }
+}
+
+// MARK: - Hashable
+
+extension Subspace: Hashable {
+    public func hash(into hasher: inout Hasher) {
+        hasher.combine(prefix)
+    }
+}
```
Comment (glbrntt) [inline Sources/FoundationDB/Subspace.swift]: Can the compiler not synthesise these conformances for you?

File: Sources/FoundationDB/Subspace.swift [hunk 3]
```diff
@@ -0,0 +1,424 @@
...
+    public var description: String {
+        let hexString = prefix.map { String(format: "%02x", $0) }.joined()
+        return "Subspace(prefix: \(hexString))"
+    }
+}
+
+// MARK: - SubspaceError
+
+/// Errors that can occur in Subspace operations
+public enum SubspaceError: Error {
```
Comment (glbrntt) [inline Sources/FoundationDB/Subspace.swift]: `enum`s generally make for bad error types as you can't add new cases without breaking API.

The most common model is having an error code and an error message wrapped up in a struct. It's not uncommon to then attach a few other things too (like an underlying error if applicable). SwiftNIO's `FileSystemError` is a good example of this (see [here](https://github.com/apple/swift-nio/blob/main/Sources/_NIOFileSystem/FileSystemError.swift)).

File: Sources/FoundationDB/Subspace.swift [hunk 4]
```diff
@@ -0,0 +1,424 @@
...
+    /// ```swift
+    /// try [0x01, 0x02].strinc()       // → [0x01, 0x03]
+    /// try [0x01, 0xFF].strinc()       // → [0x02]
+    /// try [0x01, 0x02, 0xFF, 0xFF].strinc()  // → [0x01, 0x03]
+    /// try [0xFF, 0xFF].strinc()       // throws SubspaceError.cannotIncrementKey
+    /// try [].strinc()                 // throws SubspaceError.cannotIncrementKey
+    /// ```
+    ///
+    /// - SeeAlso: `Subspace.prefixRange()` for usage with Subspace
+    public func strinc() throws -> FDB.Bytes {
```
Comment (glbrntt) [inline Sources/FoundationDB/Subspace.swift]: I just want to call out again that this is another reason we're pretty strongly opposed to using typealiases in public APIs: any extension on the typealias pollutes the aliased type.

File: Sources/FoundationDB/Subspace.swift [hunk 5]
```diff
@@ -0,0 +1,424 @@
...
+        // Strip trailing 0xFF bytes
+        var result = self
+        while result.last == 0xFF {
+            result.removeLast()
+        }
+
+        // Check if result is empty (input was empty or all 0xFF)
+        guard !result.isEmpty else {
```
Comment (glbrntt) [inline Sources/FoundationDB/Subspace.swift]: nit: `guard !X` is less readable than `if X`

File: Sources/FoundationDB/Subspace.swift [hunk 6]
```diff
@@ -0,0 +1,424 @@
...
+    /// try [0x01, 0xFF].strinc()       // → [0x02]
+    /// try [0x01, 0x02, 0xFF, 0xFF].strinc()  // → [0x01, 0x03]
+    /// try [0xFF, 0xFF].strinc()       // throws SubspaceError.cannotIncrementKey
+    /// try [].strinc()                 // throws SubspaceError.cannotIncrementKey
+    /// ```
+    ///
+    /// - SeeAlso: `Subspace.prefixRange()` for usage with Subspace
+    public func strinc() throws -> FDB.Bytes {
+        // Strip trailing 0xFF bytes
+        var result = self
```
Comment (glbrntt) [inline Sources/FoundationDB/Subspace.swift]: Given we always `mutate` self it'd make more sense to have a `mutating` version of this function and the non-mutating version on top of it. This allows a caller holding a mutable `FDB.Bytes` to avoid the unnecessary copy and allocation.

```swift
public mutating func formStrinc() throws { ... }

public func strinc() throws -> FDB.Bytes { 
  var copy = self
  copy.formString()
  return copy
}
```

File: Sources/FoundationDB/Tuple+Versionstamp.swift [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+        // Currently defaults to API 520+ behavior (4-byte offset)
+        // API < 520 used 2-byte offset, but is no longer supported
+
+        // API >= 520: Use 4-byte offset (uint32, little-endian)
+        guard position <= UInt32.max else {
+            throw TupleError.invalidEncoding
+        }
+
+        let offset = UInt32(position)
+        packed.append(contentsOf: withUnsafeBytes(of: offset.littleEndian) { Array($0) })
```
Comment (glbrntt) [inline Sources/FoundationDB/Tuple+Versionstamp.swift]: You've got an avoidable temporary array allocation here. I'd recommend adding an internal extension on `[UInt8]` to manually append the bytes of the `UInt32` to avoid going via `Array` here.

File: Sources/FoundationDB/Tuple+Versionstamp.swift [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+
+    /// Count incomplete versionstamps in tuple
+    /// - Returns: Number of incomplete versionstamps
+    public func countIncompleteVersionstamps() -> Int {
+        return elements.reduce(0) { count, element in
+            if let vs = element as? Versionstamp, !vs.isComplete {
+                return count + 1
+            }
+            return count
+        }
```
Comment (glbrntt) [inline Sources/FoundationDB/Tuple+Versionstamp.swift]: There's a `count(where:)` method you can use for this:

```swift
return elements.count { 
    if let vs = element as? Versionstamp {
        return !vs.isComplete
    }
    return false
}
```

File: Sources/FoundationDB/Tuple+Versionstamp.swift [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+    }
+
+    /// Validate tuple for use with packWithVersionstamp()
+    /// - Throws: `TupleError.invalidEncoding` if validation fails
+    public func validateForVersionstamp() throws {
+        let incompleteCount = countIncompleteVersionstamps()
+
+        guard incompleteCount == 1 else {
+            throw TupleError.invalidEncoding
+        }
```
Comment (glbrntt) [inline Sources/FoundationDB/Tuple+Versionstamp.swift]: `guard` is great for early returns. The `else` branch of the `guard` here is no earlier than the guarded path; `if` would be more idiomatic and easier to read here:

```swift
if incompleteCount != 1 {
    throw TupleError.invalidEncoding
}
```

File: Sources/FoundationDB/Tuple+Versionstamp.swift [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+        var offset = 0
+
+        while offset < bytes.count {
+            guard offset < bytes.count else { break }
+
+            let typeCode = bytes[offset]
+            offset += 1
+
+            switch typeCode {
+            case TupleTypeCode.versionstamp.rawValue:
```
Comment (glbrntt) [inline Sources/FoundationDB/Tuple+Versionstamp.swift:153]: It would be more idiomatic to turn your `typeCode` into a typed `TupleTypeCode` so that you can switch over it more naturally.

File: Sources/FoundationDB/Versionstamp.swift [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+        return transactionVersion != nil
+    }
+
+    /// Convert to 12-byte representation
+    /// - Returns: 12-byte array (10 bytes transaction version + 2 bytes user version, big-endian)
+    public func toBytes() -> FDB.Bytes {
+        var bytes = transactionVersion ?? Self.incompletePlaceholder
+
+        // User version is stored as big-endian
+        bytes.append(contentsOf: withUnsafeBytes(of: userVersion.bigEndian) { Array($0) })
```
Comment (glbrntt) [inline Sources/FoundationDB/Versionstamp.swift]: This is another avoidable array allocation.

File: Sources/FoundationDB/Versionstamp.swift [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+    public func hash(into hasher: inout Hasher) {
+        hasher.combine(transactionVersion)
+        hasher.combine(userVersion)
+    }
+
+    public static func == (lhs: Versionstamp, rhs: Versionstamp) -> Bool {
+        return lhs.transactionVersion == rhs.transactionVersion &&
+               lhs.userVersion == rhs.userVersion
+    }
+
```
Comment (glbrntt) [inline Sources/FoundationDB/Versionstamp.swift]: Any reason the compiler can't synthesise these?

File: Sources/FoundationDB/Subspace.swift [hunk 13]
```diff
@@ -0,0 +1,424 @@
...
+    /// - Parameter prefix: The binary prefix
+    ///
+    /// - SeeAlso: https://apple.github.io/foundationdb/developer-guide.html#subspaces
+    public init(prefix: FDB.Bytes) {
+        self.prefix = prefix
+    }
+
+    /// Create a subspace with a string prefix
+    /// - Parameter rootPrefix: The string prefix (will be encoded as a Tuple)
+    public init(rootPrefix: String) {
```
Comment (MMcM) [inline Sources/FoundationDB/Subspace.swift]: I am not sure about the need to privilege strings like this. In practice, I would expect that the root of a subspace would come from the Directory Layer, which we do not yet have for Swift, but should.

File: Sources/FoundationDB/Subspace.swift [hunk 14]
```diff
@@ -0,0 +1,424 @@
...
+    }
+
+    // MARK: - Key Encoding/Decoding
+
+    /// Encode a tuple into a key with this subspace's prefix
+    /// - Parameter tuple: The tuple to encode
+    /// - Returns: The encoded key with prefix
+    ///
+    /// The returned key will have the format: `[prefix][encoded tuple]`
+    public func pack(_ tuple: Tuple) -> FDB.Bytes {
```
Comment (MMcM) [inline Sources/FoundationDB/Subspace.swift:141]: I have no strong preference between `pack` / `unpack` and `encode` / `decode`, but I do not see any reason for `Subspace` and `Tuple` to differ in their choice. Perhaps others can comment on which feels more natural in this binding.

File: Sources/FoundationDB/Subspace.swift [hunk 15]
```diff
@@ -0,0 +1,424 @@
...
+    /// - Returns: A new subspace with the extended prefix
+    ///
+    /// ## Example
+    ///
+    /// ```swift
+    /// let users = Subspace(rootPrefix: "users")
+    /// let activeUsers = users.subspace("active")  // prefix = users + "active"
+    /// let userById = activeUsers.subspace(12345)  // prefix = users + "active" + 12345
+    /// ```
+    public func subspace(_ elements: any TupleElement...) -> Subspace {
```
Comment (MMcM) [inline Sources/FoundationDB/Subspace.swift:109]: Some other bindings also overload `subscript` to do this. For example, Python adds a `__getitem__`. Although Rust's `Subspace` does not implement `Index`, so it is not universal. But I have found it convenient.

**Review by MMcM — CHANGES_REQUESTED** [review]

**Comment by 1amageek** [comment]
## pack/unpack Naming Decision

Thank you for raising this important design question. After careful consideration, I've decided to **standardize on `pack/unpack` throughout the Swift bindings**, aligning with all other official FoundationDB language bindings.

### Current State Analysis

The Swift bindings initially had inconsistent naming:
- **Tuple**: `encode()` / `decode()` 
- **Subspace**: `pack()` / `unpack()` ✅
- **Versionstamp**: `packWithVersionstamp()` ✅

This inconsistency needed resolution.

### Cross-Language Consistency

All official FoundationDB bindings use **pack/unpack**:

**Python:**
\`\`\`python
packed = fdb.tuple.pack(('user', 123))
items = fdb.tuple.unpack(packed)
packed = fdb.tuple.pack_with_versionstamp(('user', vs))
\`\`\`

**Java:**
\`\`\`java
byte[] packed = Tuple.from("user", 123).pack();
Tuple tuple = Tuple.fromBytes(packed);
\`\`\`

**Go:**
\`\`\`go
packed := tuple.Tuple{"user", 123}.Pack()
elements, _ := tuple.Unpack(packed)
\`\`\`

**C++** (via bindings documentation): Uses "pack" terminology

### Why \`pack/unpack\` Instead of \`encode/decode\`

#### 1. **Established FDB Terminology**
- Official documentation: "tuple packing"
- Community discussions: "pack your data"
- Existing APIs: \`packWithVersionstamp()\`

#### 2. **Avoids Confusion with Swift's Codable**
Swift developers strongly associate \`encode/decode\` with the \`Codable\` protocol:
\`\`\`swift
// Swift's Codable
let data = try JSONEncoder().encode(user)
let user = try JSONDecoder().decode(User.self, from: data)
\`\`\`

Using \`pack/unpack\` clearly signals this is **FDB-specific tuple encoding**, not general Swift serialization.

#### 3. **Semantic Clarity**
- **pack**: "Compress multiple elements into ordered bytes" (with ordering preservation)
- **encode**: "Convert to bytes" (generic serialization)

Tuple packing is special—it preserves lexicographic ordering, which is critical for range queries.

#### 4. **Internal vs Public API**
The \`TupleElement\` protocol still uses \`encodeTuple()\`/\`decodeTuple()\` internally (for implementation), while the public API uses \`pack()\`/\`unpack()\` (for users). This separation is appropriate:
\`\`\`swift
// Public API - user-facing
let bytes = tuple.pack()
let elements = try Tuple.unpack(from: bytes)

// Internal protocol - implementation detail
protocol TupleElement {
    func encodeTuple() -> FDB.Bytes
    static func decodeTuple(from: FDB.Bytes, at: inout Int) throws -> Self
}
\`\`\`

### Implementation

I've updated the code to use \`pack/unpack\` consistently:
- \`Tuple.encode()\` → \`Tuple.pack()\`
- \`Tuple.decode(from:)\` → \`Tuple.unpack(from:)\`
- Subspace already uses \`pack/unpack\` ✅
- \`packWithVersionstamp()\` unchanged ✅

All 150 tests pass with the new API.

### Benefits

1. **Cross-language consistency**: Developers moving between Python/Java/Go/Swift will use the same terminology
2. **Clear FDB context**: No confusion with Swift's Codable
3. **Better searchability**: Searching "FoundationDB pack" yields relevant results across all languages
4. **Matches existing APIs**: \`packWithVersionstamp()\` already established this pattern

This decision prioritizes **FoundationDB ecosystem consistency** over Swift-specific conventions, which is appropriate for a database client library where cross-language compatibility is valuable.

**Comment by 1amageek** [comment]
@glbrntt @MMcM Thank you both very much for your thorough and insightful code reviews! Your feedback has significantly improved the code quality, Swift idiomaticity, and API design of this PR.

### Changes Implemented

**@glbrntt's review comments:**

1. ✅ **Generic Collection for decode** - Noted as valuable future optimization (not included in this PR to keep scope manageable)
2. ✅ **Compiler synthesis for Equatable/Hashable** - Removed manual implementations
3. ✅ **Struct-based SubspaceError** - Converted to SwiftNIO-style error model for extensibility
4. ✅ **Avoid extending FDB.Bytes typealias** - Moved to `FDB.strinc()` static method
5. ✅ **Guard vs if improvements** - Replaced unnecessary guard statements
6. ✅ **Use count(where:)** - Replaced reduce with more idiomatic approach
7. ✅ **TupleTypeCode enum in switch** - Added type-safe enum conversion
8. ✅ **Optimize withUnsafeBytes** - Eliminated unnecessary allocations

**@MMcM's review comments:**

1. ✅ **Remove rootPrefix initializer** - Removed string initializer; prefixes should come from Directory Layer (next PR)
2. ✅ **Standardize pack/unpack naming** - Changed `Tuple.encode()/decode()` to `pack()/unpack()` for cross-language consistency (detailed rationale in separate comment)
3. ✅ **Add subscript operator** - Implemented for convenient nested subspace creation, matching Python's `__getitem__` pattern

All 150 tests pass successfully after these changes.

### Ready for Re-review

I've committed these changes in separate, focused commits for easier review:
- Compiler synthesis improvements
- Code quality enhancements  
- Error model refactoring
- API namespace cleanup
- Naming alignment with other FoundationDB bindings
- Test updates

The code is now more idiomatic Swift while maintaining strong cross-language compatibility.

**Note**: I'm not a native English speaker and am using AI assistance to help communicate my intent clearly. Please let me know if anything needs clarification.

Thank you again for your time and expertise! I look forward to your feedback.
